### PR TITLE
iRprop- local optimiser.

### PR DESCRIPTION
**Pull request by MichaelClerx**
See #1105

Local gradient using optimiser with adaptive step size.
Seems to work pretty well

https://doi.org/10.1016/S0925-2312(01)00700-7

The easiest to read psuedo code for this algorithm is in a figure in the paper above.
Note that the `< 0` / `> 0` is deliberate: the step size is not updated for `= 0`.

### CODE REVIEW

File: pints/_optimisers/_irpropmin.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        1. It sets the update at this iteration to zero (using
+           ``sign(df_j[i]) * step_size_j[i] = 0 * step_size_j[i]``).
+        2. It ensures that the next iteration is performed (since
+           ``df_j[i + 1] * df_j[i] = 0`` so neither if statement holds).
+
+    In this implementation, the ``step_size`` is initialised as ``sigma_0``,
+    the increase (0.5) & decrease factors (1.2) are fixed, and a minimum step
+    size of ``1e-3 * min(sigma0)`` is enforced.
+
+    This is an unbounded methods: Any ``boundaries`` will be ignored.
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:45]: typo: methods -> method

File: pints/_optimisers/_irpropmin.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        # Adapt step sizes
+        self._step_size[dprod > 0] *= self._eta_max
+        self._step_size[dprod < 0] *= self._eta_min
+
+        # Bound step sizes
+        if self._step_min is not None:
+            self._step_size = np.maximum(self._step_size, self._step_min)
+        #if self._step_max is not None:
+        #    self._step_size = np.minimum(self._step_size, self._step_max)
+
```
Comment (DavAug) [inline pints/_optimisers/_irpropmin.py:168]: Do we want to keep this, even if we introduce no upper bound?

**Review by DavAug — APPROVED** [review]
Looks very good to me, @MichaelClerx 
I agree with not exposing the hyperparameters for now, given that according to the literature the algorithm appears to be quite robust to changes in them.

I just have minor comments. Happy for this to be merged ☺️

**Comment by MichaelClerx** [comment]
Thanks @DavAug !

**Pull request merged**